### PR TITLE
Make the mock attribute storage implementation configurable at runtime

### PR DESCRIPTION
**Pull request by ksperling-apple**

### CODE REVIEW

File: src/app/util/mock/MockNodeConfig.cpp [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+    return findById(attributes, attributeId, outIndex);
+}
+
+MockEndpointConfig::MockEndpointConfig(EndpointId aId, std::initializer_list<MockClusterConfig> aClusters) :
+    id(aId), clusters(aClusters), mEmberEndpoint{}
+{
+    VerifyOrDie(aClusters.size() < UINT8_MAX);
+
+    // Note: We're copying all the EmberAfClusters because they need to be contiguous in memory
+    for (auto && cluster : clusters)
```
Comment (tcarmelveilleux) [inline src/app/util/mock/MockNodeConfig.cpp:86]: Curious why it needs to be && rather than &.

File: src/app/util/mock/Functions.h [hunk 2]
```diff
@@ -26,15 +26,24 @@
...
 namespace chip {
 namespace Test {
+
 CHIP_ERROR ReadSingleMockClusterData(FabricIndex aAccessingFabricIndex, const app::ConcreteAttributePath & aPath,
                                      app::AttributeReportIBs::Builder & aAttributeReports,
                                      app::AttributeValueEncoder::AttributeEncodeState * apEncoderState);
 void BumpVersion();
 DataVersion GetVersion();
+
+/// Configures the mock attribute storage to use the specified configuration.
```
Comment (tcarmelveilleux) [inline src/app/util/mock/Functions.h:42]:
```suggestion
/// Configures the singular global mock attribute storage to use the specified configuration.
```

**Review by tcarmelveilleux — APPROVED** [review]

File: src/app/util/mock/MockNodeConfig.cpp (hunk 1 above)
Comment (ksperling-apple) [reply src/app/util/mock/MockNodeConfig.cpp:86]: `auto const &` is probably better... I was under the impression that `auto &&` is a good default for loops but seems like that's not necessarily the case.

File: src/app/util/mock/MockNodeConfig.h [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+/**
+ *    @file
+ *     Helpers for configuring the mocked attribute-storage.cpp
```
Comment (andy31415) [inline src/app/util/mock/MockNodeConfig.h:21]: please delete this file comment.

File: src/app/util/mock/attribute-storage.cpp [hunk 4]
```diff
@@ -312,42 +266,24 @@ uint16_t emberAfLongStringLength(const uint8_t * buffer)
...
-    if (index == kEmberInvalidEndpointIndex)
-    {
-        return false;
-    }
-
-    return clusterId; // Mock version return true as long as the endpoint is
-                      // valid
+    auto config = GetMockNodeConfig();
+    VerifyOrReturnValue(index < config.endpoints.size(), false);
+    return true; // TODO: TestSceneTable relies on returning true here, we should fix (or use a dedicated) mock config instead
```
Comment (andy31415) [inline src/app/util/mock/attribute-storage.cpp:271]: could we add a github issue and reference it here?

**Review by andy31415 — APPROVED** [review]

**Review by jmartinez-silabs — APPROVED** [review]

**Pull request merged**